src/python/pants/engine/internals/rule_visitor.py

```diff
@@ -299,9 +299,10 @@ def visit_Call(self, call_node: ast.Call) -> None:
                 self.awaitables.append(
                     self._get_legacy_awaitable(call_node, is_effect=issubclass(func, Effect))
                 )
-            elif inspect.isfunction(func) and (rule := getattr(func, "rule", None)) is not None:
+            elif (
+                inspect.isfunction(func) and (rule_id := getattr(func, "rule_id", None)) is not None
+            ):
                 # Is a direct `@rule` call.
-                rule_id = rule.canonical_name
```

Comment:
This was all we needed the rule instance for, so we can just pass it in directly, before we create the rule instance.